### PR TITLE
HIP-513 for traceability extension

### PR DESCRIPTION
**Pull request by mustafauzunn**
**Description**:

Remove ContractStateChange from ContractFunctionResult
Add new sidecar record types 

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: HIP/hip-513.md [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+## Rationale
+
+[HIP-435](./hip-435.md) allows the creation of sidecar record files containing
+detailed information about transaction executions without bloating the main
+transaction record stream. This allows us to externalise verbose debug information
+about smart contract transactions that was previously expensive to download and
+process every time.
+
+We will introduce a couple of sidecar record types, one containing the minimum
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+last-call-date-time:
+created: 2022-06-08
+discussions-to:
+updated:
+requires: 206, 435, 260
+---
```
Comment (shemnon) [inline HIP/hip-513.md:14]: This replaces HIP-260 so we should say so in the metadata.

```suggestion
replaces: 260
---
```

When it goes to accepted we will also need to update HIP-260 with "superseded-by" metadata

File: HIP/hip-513.md [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    Precompile = 3;
+
+    /**
+     * Call, but to system contract like HTS or ERC20 facades over Token accounts
+     */
+    System = 4;
+}
+
+/**
+ * A finder grained action with a function result. Sometimes called "internal transactions." The function call itself
```
Comment (shemnon) [inline HIP/hip-513.md:111]: s/finer/finder/

File: HIP/hip-513.md [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        ContractID recipientContract = 7;
+
+        /**
+         * If the action was directed at an invalid solidity address, what that address was.
+         */
+        bytes invalidSolidityAddress = 8;
+    }
+
+    /**
+     * The value (int tinybars) that is associated with this action.
```
Comment (shemnon) [inline HIP/hip-513.md:169]: s/int/in/

File: HIP/hip-513.md [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    repeated ContractID createdContractIDs = 7;
+    
+    // field to remove
+    repeated ContractStateChange stateChanges = 8;
+}
+```
+
+## Backwards Compatibility
+
+This HIP changes ContractFunctionResult protobuf by removing ContractStateChange from it.
```
Comment (shemnon) [inline HIP/hip-513.md:259]: We should discuss the impact of removing a proto field.  In the event it was populated it will be ignored by future processing, but will not break the validity of mesasges containing it..  IIRC previewnet and testnet had this enabled.

**Review by shemnon — CHANGES_REQUESTED** [review]
metadata and reserved field are must fix.

File: HIP/hip-513.md [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+  repeated ContractStateChange contract_state_changes = 1;
+}
+```
+
+For each slot read or written by the smart contract transaction
+a `ContactStateChange` record should be added to the `ContractStateChanges`.
+
+### Add ContractAction sidecar record type
+
+Add and populate a new sidecar record of type `ContractAction`. It will
```
Comment (steven-sheehy) [inline HIP/hip-513.md:73]: Shouldn't there be a `ContractActions` with a repeated `ContractAction`? How would you show multiple method calls for a particular contract call transaction?

File: HIP/hip-513.md [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+    /**
+     * CREATE, CREATE2, and first action of ContractCreate.
+     */
+
+    Create = 2;
+
+    /**
+     * like Call, but to precompiled contracts (0x1 to 0x9 as of Berlin)
+     */
+    Precompile = 3;
```
Comment (steven-sheehy) [inline HIP/hip-513.md:102]: Precompiles show up in the record stream as a separate HAPI transaction with a unique consensus timestamp. Will the contract actions for a precompile be associated with the parent contractcall/contractcreate/ethereumtransaction timestamp or with its child precompile transaction timestamp?

File: HIP/hip-513.md [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    uint32 callDepth = 15;
+}
+```
+
+For every smart contract transaction a `ContractAction` record should be added to the sidecar record.
+
+### Add ContractBytecode sidecar record type
+
+Add and populate a new sidecar record of type `ContractBytecode`. It will
+have two fields - `deploymentBytecode` and `runtimeBytecode`. Fields will
```
Comment (steven-sheehy) [inline HIP/hip-513.md:216]: Will this finally allow child contract creates to send their init code and runtime bytecode? And if so the timestamp associated with that be will be for the child transaction, right? Can we call this scenario out here.

File: HIP/hip-513.md [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+allow tracking smart contracts.
+
+```
+message ContractBytecode {
+
+    /**
+    * Contract bytecode during deployment
+    */
+    bytes deploymentBytecode = 1;
```
Comment (steven-sheehy) [inline HIP/hip-513.md:225]: What scenarios will this be populated? Will it be populated even `ContractCreateTransaction.initcode` is populated? Only if ContractCreationTransaction.fileID` is populated? Only for child creates?

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

**Comment by steven-sheehy** [comment]
If a migration of traceability state to mirror node is needed then this process should be outlined in this document. Particularly what timestamp would be associated with this migration and how to differentiate between migration and non migration data

File: HIP/hip-513.md [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+last-call-date-time:
+created: 2022-06-08
+discussions-to:
+updated:
+requires: 206, 435, 260
+---
+
+## Abstract
+
+This HIP overrides [HIP-260](./hip-260.md) by moving traceability info from `ContractFunctionResult` to sidecar.
```
Comment (Nana-EC) [inline HIP/hip-513.md:18]: There's an assumption of sidecar here. The term is an overloaded one and also ambiguous depending on experience.
Is there a link that can provided or a quick sentence to give context on what that is for community readers.

File: HIP/hip-513.md [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+Extends traceability with new sidecar records `ContractAction` and `ContractBytecode`.
+
+Proposes additional data placed in sidecar transaction records to enable full
+replay of transactions by mirror nodes and downstream users of Hedera.
+
+## Motivation
+
+Enabling mirror nodes to download traceability info only if they need or want to process it.
+
+Enabling full replay of smart contract service transactions will improve the
```
Comment (Nana-EC) [inline HIP/hip-513.md:29]: The flow will also support estimation of execution or values like gas to be consumed.
So it's worth noting that.

File: HIP/hip-513.md (hunk 10 above)
Comment (Nana-EC) [reply HIP/hip-513.md:18]: You can reference [HIP-435](./hip-435.md) here

File: HIP/hip-513.md (hunk 9 above)
Comment (Nana-EC) [reply HIP/hip-513.md:225]: This should be populated for every contract.
Its the bytecode that's actually stored on chain, where as the previous contains constructor code and the runtime bytecode

File: HIP/hip-513.md [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+For every smart contract transaction a `ContractAction` record should be added to the sidecar record.
+
+### Add ContractBytecode sidecar record type
+
+Add and populate a new sidecar record of type `ContractBytecode`. It will
+have two fields - `deploymentBytecode` and `runtimeBytecode`. Fields will
+allow tracking smart contracts.
+
+```
+message ContractBytecode {
```
Comment (Nana-EC) [inline HIP/hip-513.md:220]: This might actually be a source for confusion.
In other sectors of the ecosystem these are conventionally named bytecode and deployedBytecode i.e.
```proto
bytes bytecode = 1;
bytes deployedBytecode = 2;
```

File: HIP/hip-513.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-513.md:225]: Yeah. I'd actually prefer conforming with industry and just setting this to `bytecode` per my suggestion above

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Looking good, some suggestion and discussion items to agree on

File: HIP/hip-513.md [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+    // existing fields
+    ContractID contractID = 1;
+    bytes contractCallResult = 2;
+    string errorMessage = 3;
+    bytes bloom = 4;
+    uint64 gasUsed = 5;
+    repeated ContractLoginfo logInfo = 6;
+    repeated ContractID createdContractIDs = 7;
+    
+    // field to remove
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md:260]: DONE

File: HIP/hip-513.md (hunk 3 above)
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md (hunk 4 above)
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+author: Stoyan Panayotov <stoyan.panayotov@limechain.tech>, Mustafa Uzun <mustafa.uzun@limechain.tech>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+last-call-date-time:
+created: 2022-06-08
+discussions-to:
+updated:
+requires: 206, 435, 260
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md:13]: DONE

File: HIP/hip-513.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+        /**
+         * The transaction itself failed without an explicit `REVERT`
+         */
+        bytes error = 14;
+    }
+
+    /**
+     * The nesting depth of this call. The original action is at depth=0.
+     */
+    uint32 callDepth = 15;
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+
+```
+/**
+ * The type of action described by the action proto.
+ */
+enum ContractActionType {
+    /**
+     * default non-value.
+     */
+    NoAction = 0;
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+        /**
+         * If the caller was a smart contract account, the ContractID.
+         */
+        ContractID callingContract = 3;
+    }
+
+    /**
+     * Gas given to this action, the upper limit of gas this action can spend.
+     */
+    uint64 gas = 4;
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md (hunk 6 above)
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md (hunk 11 above)
Comment (mustafauzunn) [reply HIP/hip-513.md:31]: DONE

File: HIP/hip-513.md (hunk 12 above)
Comment (mustafauzunn) [reply HIP/hip-513.md:228]: DONE

File: HIP/hip-513.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply HIP/hip-513.md]: Thanks. 
If Steven feels strongly on `init_bytecode` I'm okay with that as I've seen initcode and bytecode used interchangeably.

Comment (steven-sheehy) [reply HIP/hip-513.md]: My mistake, this comment was intended to be on the init bytecode field. So with that in mind, my comments are still applicable.

File: HIP/hip-513.md (hunk 7 above)
Comment (steven-sheehy) [reply HIP/hip-513.md]: Ok, so HTS precompiled contracts won't show up separately in the sidecar but will always be attached to a top level HAPI transaction list of actions? That makes sense to me. Might be good to clarify what sorts of transactions will have actions in the sidecar in the HIP.

File: HIP/hip-513.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (shemnon) [reply HIP/hip-513.md]: Etehreum standards refer to initcode, such as "EIP-3860: Limit and meter initcode".  The EIPs have 17 instances of "init code" and 120 instances of "initcode," so the single-word form is most common.

File: HIP/hip-513.md (hunk 7 above)
Comment (shemnon) [reply HIP/hip-513.md]: I would prefer that precompiles still create contract actions in the sidecars.  (a) we would have to detect the system contract and _not_ emit a contract action and (b) if we took it out we would need to "swizzle" the data back in for a trace of the call.

I would rather not take effort to take it out and put it back in to save data in the sidecar.

File: HIP/hip-513.md [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+
+    /**
+    * Contract bytecode during deployment
+    */
+    bytes bytecode = 1;
+    
+    /**
+    * Contract bytecode after deployment
+    */
+    bytes deployed_bytecode = 2;
```
Comment (shemnon) [inline HIP/hip-513.md:237]: consider `runtime_bytecode`, it is used. times in EIP specs vs 1 instance of "deployed bytecode"

File: HIP/hip-513.md [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+about smart contract transactions that was previously expensive to download and
+process every time.
+
+We will introduce a couple of sidecar record types, one containing the minimum
+needed data that allows a user to fully replay the transaction and derive the same
+set of results, and the other containing an entire operation-by-operation trace.
+
+## User stories
+
+As a user of a mirror node I want to be able to see a list of state changes, a
```
[the remark this answers is not in the document]
Comment (stoqnkpL) [reply HIP/hip-513.md]: > as be able to execute a contract against a mirror node with a contract's actual state without modifying state or charging gas.
Will this be specified as part of this HIP? I think this behavior should be described in a separate HIP, being able to execute transactions against a mirror node is quite a big change/feature. Or do you mean to just mention it here and maybe reference that other HIP when it's written?

File: HIP/hip-513.md [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+
+    /**
+     * The actual gas spent by this action.
+     */
+    uint64 gasUsed = 10;
+
+    /**
+     * Was the action successful?
+     */
+    bool success = 11;
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md:307]: DONE

File: HIP/hip-513.md (hunk 18 above)
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md [hunk 21]
```diff
@@ -0,0 +1,342 @@
...
+message ContractBytecodes {
+    repeated ContractBytecode contract_bytecodes = 1;
+}
+
+message ContractBytecode {
+
+    /**
+    * Contract bytecode during deployment
+    */
+    bytes bytecode = 1;
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md (hunk 7 above)
Comment (stoqnkpL) [reply HIP/hip-513.md]: Precompiles will create `ContractActions` in _a_ sidecar record, the question is what timestamp that sidecar should have (which transaction will the sidecar be associated with), the top-level tx, or the child record?

Comment (stoqnkpL) [reply HIP/hip-513.md]: For example, would it make sense to have a Child record of type `CryptoApprove` for some HTS precompile invocation, and have a sidecar record with `ContractActions` that is associated with the child record instead of the parent ContractCall transaction

File: HIP/hip-513.md (hunk 8 above)
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md (hunk 9 above)
Comment (mustafauzunn) [reply HIP/hip-513.md]: Good point. We can populate initcode only if `ContractCreationTransaction.fileID` is populated
or we have child contract create.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]
Looking good. Still need to clarify the migration aspect.

File: HIP/hip-513.md [hunk 22]
```diff
@@ -0,0 +1,342 @@
...
+
+    /**
+     * The value (in tinybars) that is associated with this action.
+     */
+    int64 value = 9;
+
+    /**
+     * The actual gas spent by this action.
+     */
+    uint64 gas_used = 10;
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

File: HIP/hip-513.md [hunk 23]
```diff
@@ -0,0 +1,340 @@
...
+
+message ContractActions {
+    repeated ContractAction contract_actions = 1;
+}
+
+/**
+ * A finer grained action with a function result. Sometimes called "internal transactions." The function call itself
+ * will be the first action in a list, followed by sub-action in the order they were executed.
+ */
+message ContractAction {
```
Comment (steven-sheehy) [inline HIP/hip-513.md:126]: What will uniquely identify a contract action? The consensus timestamp plus what additional field(s)? Shouldn't there be a slot key to help identify or should we just use the order of the actions in the list as an index so the primary key is `(consensus_timestamp, index)`?

Comment (shemnon) [reply HIP/hip-513.md:126]: Ordinal nested order is the industry standard - https://openethereum.github.io/JSONRPC-trace-module#traceaddress-field.

We can reconstruct this index with the order of the contract actions and the `call_depth` attribute.  We can also simply do order of call within the transaction regardless of depth.  Array == nested ID integer = overall order.  We would also need the parent transaction ID.

File: HIP/hip-513.md (hunk 7 above)
Comment (stoqnkpL) [reply HIP/hip-513.md]: As agreed on the parking lot yesterday, we will create `ContractAction` entries and use the top-level record timestamp in the `TransactionSidecarRecord` where we list them.

File: HIP/hip-513.md [hunk 24]
```diff
@@ -87,32 +87,27 @@ Which will improve debugging and auditing.
  * The type of action described by the action proto.
  */
 enum ContractActionType {
-    /**
-     * default non-value.
-     */
-    NO_ACTION = 0;
```
Comment (shemnon) [inline HIP/hip-513.md]: Please undo this change.  If NO_ACTION is not desired then it can be changed to UNKNOWN.  Zero values are elided from the wire and I don't want the meaning of the value incorrectly implied by an incomplete serialization.

This is also Google's recommended practice - https://developers.google.com/protocol-buffers/docs/style#enums

**Review by shemnon — CHANGES_REQUESTED** [review]
un-do contract action type renumbering

File: HIP/hip-513.md (hunk 24 above)
Comment (mustafauzunn) [reply HIP/hip-513.md]: [Here](https://github.com/hashgraph/hedera-protobufs/pull/211#discussion_r903857306) @steven-sheehy  suggested to change this enums, because they are populated by consensus nodes and just adds complexity in processing it since it's unclear how the mirror node should react to an UNKNOWN. On other hand I agree with you that enums with UNKNOWN is Google's recommended practice. Maybe we need to discuss which form to use

Comment (mustafauzunn) [reply HIP/hip-513.md]: DONE

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Review by shemnon — APPROVED** [review]

**Review by steven-sheehy — APPROVED** [review]
